### ISSUE TITLE
More flexibility to validation set selection

### ISSUE THREAD

**Issue by arasmus** [mila-udem/fuel#101]
Currently I don't see any clean way to do traditional k-fold or leave-one-out cross-validation or any more flexible division of the training data to the validation set. There is only this (start, end) pair available and it is not flexible enough for this purpose (not possible to define two ranges for training and leave in-between samples for validation). This should be a very typical approach in machine learning in general, but with DL datasets are so big that people don't afford to do this. Still, I'd like fuel to support it. What do you think?

I would be personally happy with just adding a shuffle switch that would randomize the order of samples before division. But of course it would need to be internally randomized only once when opened to avoid train and validation from overlapping.

So this is what I traditionally do:
```
train = MNIST("train", end=50000) # first 50k samples
valid = MNIST("train", start=50000) # the last 10k samples
test = MNIST("test")
```

One work-around would be to only create one dataset from "train" and use Transformers somehow, but this data division is so important part of the process that I think fuel should support it directly. Especially because it is pretty easy to implement and support.

**Comment by bartvm** [comment]
How would you want to use it exactly? Because it seems to me like your support for lists as arguments to `BatchScheme` supports exactly this case. You just need to look at the different validation sets as different ways of *iterating* over your data, instead of changing the data you load. So for k-fold cross-validation just do:

```python
mnist = MNIST('train')
data_streams = []
total_set = numpy.arange(50000)
for validation_set in numpy.split(total_set, 20):
    training_set = numpy.setdiff1d(total_set, validation_set)
    data_streams += [DataStream(mnist, iteration_scheme=ShuffledScheme(training_set), 128)]
```

**Comment by vdumoulin** [comment]
I like @bartvm's idea of using iteration schemes a lot. (In fact, it might be a good idea to push that idea forward in a tutorial in the online documentation.)

On the other hand, being able to specify a split as a list of disjoint indices is something that we have in mind, but we haven't had the time to implement that yet. I just created an issue (#102) to make that more official.

**Comment by bartvm** [comment]
Right, we do need support for splits as non-contiguous lists of indices, but that is only for cases where there is some sort of pre-defined split that should be made part of the dataset. @arasmus seems to be talking more about things like bootstrap aggregating, k-fold cross-validation, etc. where you create the validation sets yourself. In those cases I think it's better not to see them as splits belonging to the dataset, but as iteration schemes belonging to the data stream.

**Comment by vdumoulin** [comment]
I don't necessarily disagree with you, but we already have a form of support for creating arbitrary validation sets in the `subset` argument: within the context of the split, `subset` defines a subset of the data that is used, and in that sense, it could also support a list of indices that represent a subset of the data *within the context of the split*.

Like I said, I like your idea of creating a subset of an official split using iterators, because it simplifies implementation a great deal by making sure that irrelevant examples are never requested in the first place instead of pretending they don't exist.

The one place where I could see a use case for having `subset` be a list of indices is if the user wants to load the subset into memory: while the split itself may be too big, the subset may fit into memory, and the user may want that.

**Comment by bartvm** [comment]
Well, there's another use-case I mentioned before: Datasets that are a mix of multiple datasets e.g. text corpora (WMT, billion words) are actually a mix of a variety of datasets (UN, Europarl, News Commentary, etc.) In the case of billion words these are then shuffled before split into train and test sets. I can imagine use cases where you would still want to retrieve all the examples from one of the original corpora, so that you can see how your model performs across different domains.

Just like datasets support a `start` and `stop` argument, we could technically support a `subset` argument that takes a list of indices. But still, to do k-fold cross validation I wouldn't create 10 datasets that load different subsets. I would load a single dataset with the union of those indices, and then iterate over that one dataset in 10 different ways.

**Comment by vdumoulin** [comment]
Right, it's a matter of use case patterns. Like I said, that should be documented as "patterns for best practices" in the online documentation when we have time to do it.

**Comment by arasmus** [comment]
> I think it's better not to see them as splits belonging to the dataset, but as iteration schemes belonging to the data stream.

@bartvm, thanks, maybe doing the valid/train split outside of Dataset is not that bad after all. I'll follow your example. I was thinking that maybe dataset should provide this as a service if this becomes a common pattern but this is a design choice. I agree with vdumoulin that maybe it could be at least part of a tutorial.

Feel free to close this issue.
